# Supervised Learning Models

In this section, multiple classification models are trained to predict whether a patient has heart disease.

The models used are:

- Logistic Regression
- Support Vector Machine (SVM)
- Decision Tree
- Random Forest
- K-Nearest Neighbors (KNN)

Each model is built using a Scikit-learn Pipeline that combines preprocessing and model training.

GridSearchCV is used to test different hyperparameter combinations using Cross-Validation. The F1-score is used as the scoring metric because the target variable has a moderate class imbalance.

The best hyperparameters and best Cross-Validation F1-score are recorded for each model. The models will be evaluated and compared in the Evaluation notebook.

## Load the Dataset and prepare it

In [ ]:
import pandas as pd
df= pd.read_csv(r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\data\dataprocessed_heart_data.csv")
df

In [ ]:

bool_cols = [
    "exercise_induced_angina",
    "family_history",
    "wearable_owner"
]

df[bool_cols] = df[bool_cols].astype(int)

In [ ]:
X = df.drop("has_heart_disease", axis=1)
y = df["has_heart_disease"]

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split( X, y,test_size=0.2,random_state=42)

In [5]:
cat_cols = X.select_dtypes(include="object").columns
num_cols = X.select_dtypes(exclude="object").columns
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
pre = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), num_cols),
        ("cat", OneHotEncoder(handle_unknown="ignore"), cat_cols)
    ]
)

C:\Users\Hp\AppData\Local\Temp\ipykernel_13996\3749316474.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = X.select_dtypes(include="object").columns


In [6]:
df.dtypes

age                            int64
sex                              str
resting_bp_systolic            int64
resting_bp_diastolic           int64
cholesterol_total              int64
hdl                            int64
ldl                            int64
triglycerides                  int64
fasting_blood_sugar            int64
hba1c                        float64
bmi                          float64
resting_heart_rate             int64
max_heart_rate_achieved        int64
chest_pain_type                  str
exercise_induced_angina        int64
st_depression                float64
family_history                 int64
smoker_status                    str
alcohol_units_per_week       float64
exercise_minutes_per_week      int64
sleep_hours                  float64
stress_score                 float64
wearable_owner                 int64
daily_steps                    int64
diet_quality_score           float64
has_heart_disease              int64
dtype: object

## 1. Build the Logistic Regression Pipeline

In [7]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
log_pip=Pipeline([
    ('pre',pre),
    ('model',LogisticRegression(random_state=42))
    
])
param_grid_logistic = {
    "model__C": [0.1, 1, 10,100],
    "model__penalty": ["l1", "l2"],
    "model__solver": ["liblinear"],
    "model__class_weight": [None, "balanced"]
}
log_search=GridSearchCV(
    estimator=log_pip,
    param_grid=param_grid_logistic,
    cv=5,
    scoring='f1'
)
log_search.fit(X_train,y_train)
print(log_search.best_params_)
print(log_search.best_score_)

c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\linear_model\_logistic.py:1403: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', l1_ratio set to a float between 0 and 1 instead of penalty='elasticnet', and C=np.inf instead of penalty=None.
  warnings.warn(
c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\linear_model\_logistic.py:1429: UserWarning: Inconsistent values: penalty=l1 with l1_ratio=0.0. penalty is deprecated. Please use l1_ratio only.
  warnings.warn(
c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\linear_model\_logistic.py:1403: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and u

{'model__C': 10, 'model__class_weight': None, 'model__penalty': 'l2', 'model__solver': 'liblinear'}
0.8251469001858439


c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\linear_model\_logistic.py:1403: FutureWarning: 'penalty' was deprecated in version 1.8 and will be removed in 1.10. To avoid this warning, leave 'penalty' set to its default value and use 'l1_ratio' or 'C' instead. Use l1_ratio=0 instead of penalty='l2', l1_ratio=1 instead of penalty='l1', l1_ratio set to a float between 0 and 1 instead of penalty='elasticnet', and C=np.inf instead of penalty=None.
  warnings.warn(


### Train the Baseline Model

## 2. Build the SVM Pipeline

In [8]:
from sklearn.svm import SVC
svm_pip=Pipeline([
    ('pre',pre),
    ('model',SVC(probability=True,random_state=42))
    
])

In [9]:
from sklearn.model_selection import GridSearchCV
param_grid = {
    "model__C": [0.1, 1, 10,100],
    "model__kernel": ["linear", "rbf"],
}
svm_search = GridSearchCV(
    estimator=svm_pip,
    param_grid=param_grid,
    cv=4,
    scoring="f1"
)
svm_search.fit(X_train, y_train)
print(svm_search.best_params_)
print(svm_search.best_score_)


c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(
c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(
c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(
c:\Users\Hp\AppData\Local\Programs\Python\Python314\Lib\site-packages\sklearn\svm\_base.py:239

{'model__C': 1, 'model__kernel': 'linear'}
0.8233034699272506


## 3. Decision Tree

In [10]:
from sklearn.tree import DecisionTreeClassifier
tree_pip=Pipeline([
    ('pre',pre),
    ('model',DecisionTreeClassifier(random_state=42))
])
param_grid_tree = {
    "model__max_depth": [3, 5, 10],
    "model__min_samples_split": [2, 5],
    "model__min_samples_leaf": [1, 2],
    
}
tree_search = GridSearchCV(
    estimator=tree_pip,
    param_grid=param_grid_tree,
    cv=5,
    scoring="f1"
)
tree_search.fit(X_train,y_train)
print(tree_search.best_params_)
print(tree_search.best_score_)


{'model__max_depth': 5, 'model__min_samples_leaf': 2, 'model__min_samples_split': 2}
0.7330122203929225


## 4. Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rns_pip=Pipeline([
    ('pre',pre),
    ('model',RandomForestClassifier(random_state=42))
])
param_grid_rf = {
    "model__n_estimators": [100, 200,300,400],
    "model__max_depth": [5, 10, None],
    "model__min_samples_split": [2, 5],
    "model__min_samples_leaf": [1, 2],
    "model__class_weight": [None, "balanced"]
}
rf_model=GridSearchCV(estimator=rns_pip,param_grid=param_grid_rf,cv=5,scoring='f1')
rf_model.fit(X_train,y_train)
print(rf_model.best_params_)
print(rf_model.best_score_)

{'model__class_weight': 'balanced', 'model__max_depth': None, 'model__min_samples_leaf': 2, 'model__min_samples_split': 5, 'model__n_estimators': 400}
0.8038437657614688


## 5. K-Nearest Neighbors (KNN)

In [12]:
from sklearn.neighbors import KNeighborsClassifier
knn_pip=Pipeline([
    ('pre',pre),
    ('model',KNeighborsClassifier())
])
param_grid_knn={
    "model__n_neighbors": [3, 5, 7, 9, 11,15,20],
    "model__weights": ["uniform", "distance"],
    "model__metric": ["euclidean", "manhattan"]
}
knn_search=GridSearchCV(
    estimator=knn_pip,
    param_grid=param_grid_knn,
    cv=5,
    scoring='f1'
    
)
knn_search.fit(X_train,y_train)
print(knn_search.best_params_)
print(knn_search.best_score_)

{'model__metric': 'euclidean', 'model__n_neighbors': 20, 'model__weights': 'distance'}
0.7105900033151104


# Model Tuning Results

Five classification models were trained and tuned using **GridSearchCV**:

- Logistic Regression
- Support Vector Machine (SVM)
- Decision Tree
- Random Forest
- K-Nearest Neighbors (KNN)

The hyperparameter combinations were evaluated using **Cross-Validation**, with **F1-score** used as the scoring metric.

F1-score was selected because the target variable has a moderate class imbalance, with approximately **69.7%** of patients belonging to class `0` and **30.3%** belonging to class `1`.

---

## 1. Logistic Regression

### Best Hyperparameters

```text
C = 10
class_weight = None
penalty = l2
solver = liblinear
```

### Best Cross-Validation F1-score

```text
0.8251
```

The tuned Logistic Regression model achieved an average Cross-Validation F1-score of approximately **82.51%**.

Among the tested hyperparameter combinations, the best model used **L2 regularization**, `C=10`, and the `liblinear` solver without class weighting.

---

## 2. Support Vector Machine (SVM)

### Best Hyperparameters

```text
C = 1
kernel = linear
```

### Best Cross-Validation F1-score

```text
0.8233
```

The tuned SVM model achieved an average Cross-Validation F1-score of approximately **82.33%**.

The best configuration used a **linear kernel** with `C=1`.

Its performance was very close to Logistic Regression.

---

## 3. Decision Tree

### Best Hyperparameters

```text
max_depth = 5
min_samples_leaf = 2
min_samples_split = 2
```

### Best Cross-Validation F1-score

```text
0.7330
```

The tuned Decision Tree achieved an average Cross-Validation F1-score of approximately **73.30%**.

The best tree had a maximum depth of `5`, which limits the complexity of the tree and can help reduce overfitting.

Its Cross-Validation performance was lower than Logistic Regression and SVM.

---

## 4. Random Forest

### Best Hyperparameters

```text
class_weight = balanced
max_depth = None
min_samples_leaf = 2
min_samples_split = 5
n_estimators = 400
```

### Best Cross-Validation F1-score

```text
0.8038
```

The tuned Random Forest achieved an average Cross-Validation F1-score of approximately **80.38%**.

The best configuration used **400 decision trees** and `class_weight="balanced"`.

The balanced class weighting gives more importance to the minority class during training.

Random Forest performed better than Decision Tree and KNN, but its Cross-Validation F1-score was lower than Logistic Regression and SVM.

---

## 5. K-Nearest Neighbors (KNN)

### Best Hyperparameters

```text
metric = euclidean
n_neighbors = 20
weights = distance
```

### Best Cross-Validation F1-score

```text
0.7106
```

The tuned KNN model achieved an average Cross-Validation F1-score of approximately **71.06%**.

The best configuration used `20` neighbors with **Euclidean distance**.

Using `weights="distance"` means that closer neighbors have more influence on the prediction than neighbors that are farther away.

KNN achieved the lowest Cross-Validation F1-score among the tested models.

---

# Cross-Validation Model Comparison

| Model | Best CV F1-score |
|---|---:|
| Logistic Regression | **0.8251** |
| SVM | **0.8233** |
| Random Forest | **0.8038** |
| Decision Tree | **0.7330** |
| KNN | **0.7106** |

---

## Initial Comparison

Based on the Cross-Validation F1-scores:

1. **Logistic Regression** achieved the highest score with approximately **82.51%**.
2. **SVM** was a very close second with approximately **82.33%**.
3. **Random Forest** achieved approximately **80.38%**.
4. **Decision Tree** achieved approximately **73.30%**.
5. **KNN** achieved approximately **71.06%**.

Logistic Regression currently shows the best Cross-Validation performance, although its performance is very close to SVM.

However, these Cross-Validation results should not be treated as the final model evaluation.

The tuned models will next be evaluated on the unseen **Test Set** using:

- Accuracy
- Precision
- Recall
- F1-score
- ROC-AUC
- Confusion Matrix

The final model comparison will be based on the test-set performance and the practical importance of different types of classification errors.

## Save the Trained Models

In [13]:
import joblib

joblib.dump(
    log_search.best_estimator_,
    r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\models\logistic_regression_pipeline.pkl"
)

joblib.dump(
    svm_search.best_estimator_,
    r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\models\svm_pipeline.pkl"
)

joblib.dump(
    tree_search.best_estimator_,
    r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\models\decision_tree_pipeline.pkl"
)

joblib.dump(
    rf_model.best_estimator_,
    r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\models\random_forest_pipeline.pkl"
)

joblib.dump(
    knn_search.best_estimator_,
    r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\models\knn_pipeline.pkl"
)

['C:\\Users\\Hp\\Desktop\\binx_mohammadabuhamed\\binx_mohammadabuhamed\\Cardiac Patients\\models\\knn_pipeline.pkl']

## Save the Test Dataset

In [15]:
data_test=X_test.copy()
data_test['has_heart_disease']=y_test
data_test.to_csv( r"C:\Users\Hp\Desktop\binx_mohammadabuhamed\binx_mohammadabuhamed\Cardiac Patients\data\test_data.csv",index=False)